core/include/gnuradio-4.0/Block.hpp

```diff
@@ -1337,290 +1426,119 @@ class Block : public lifecycle::StateMachine<Derived>, //
-                ports_status.in_samples = ports_status.nSamplesToNextTag == 0 ? std::min(ports_status.in_samples, std::max(ports_status.nSamplesToNextTagAfter, 1UZ))
-                                                                              : std::min(ports_status.in_samples, ports_status.nSamplesToNextTag);
-            }
-#ifdef _DEBUG
-            fmt::println("block {} - mark2 - in {} -> out {}", name, ports_status.in_samples, ports_status.out_samples);
-#endif
-            ports_status.in_samples  = std::min(ports_status.in_samples, ports_status.nSamplesToEosTag); // clamp input until EOS tag
-            ports_status.out_samples = ports_status.in_samples;
-            // TODO: recheck definitions of numerator and denominator
-            //   Provided we we define N := <numerator> and M := <denominator>, do we expect (assuming a simple in->out block) that
-            //  a) for N input samples M output samples are produced, or
-            //  b) for a given number of input samples X -> X * N/M output samples are produced.
-            //  Some of the logic conditions vary depending on whether we assume 'a)' or 'b)'.
-            const bool isEOSTagPresent = ports_status.nSamplesToEosTag == 0                          //
-                                      || ports_status.nSamplesToEosTag < ports_status.in_min_samples //
-                                      || ports_status.nSamplesToEosTag < numerator;
-#ifdef _DEBUG
-            if (isEOSTagPresent) {
-                fmt::println("##block {} received EOS tag at {} < in_min_samples {}", name, ports_status.nSamplesToEosTag, ports_status.in_min_samples);
-            }
-#endif
-
-            if (isEOSTagPresent || lifecycle::isShuttingDown(this->state())) {
+        // evaluate number of available and processable samples
+        const auto [minSyncIn, maxSyncIn, maxSyncAvailableIn, hasAsyncIn]     = getPortLimits(inputPorts<PortType::STREAM>(&self()));
+        const auto [minSyncOut, maxSyncOut, maxSyncAvailableOut, hasAsyncOut] = getPortLimits(outputPorts<PortType::STREAM>(&self()));
+        auto [hasTag, nextTag, nextEosTag, asyncEoS]                          = getTagPositions();
+        auto       maxChunk                                                   = getMergedBlockLimit(); // handle special cases for merged blocks. TODO: evaluate if/how we can get rid of these
+        const auto inputSkipBefore                                            = inputSamplesToSkipBeforeNextChunk(std::min({ maxSyncAvailableIn, nextTag, nextEosTag }));
+        const auto availableToProcess          = std::min({ maxSyncIn, maxChunk, (maxSyncAvailableIn - inputSkipBefore), (nextTag - inputSkipBefore), (nextEosTag - inputSkipBefore) });
+        const auto availableToPublish          = std::min({ maxSyncOut, maxSyncAvailableOut });
+        const auto [decimatedIn, decimatedOut] = decimate(minSyncIn, availableToProcess, minSyncOut, availableToPublish);
```

Comment:
I like the structured binding which makes the code more readable. Maybe these names could be also used in the generating methods w.r.t. naming which make those also more readable.